src/audio/aec/aec.c

```diff
@@ -0,0 +1,393 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/aec/aec.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#define AEC_BLK_LENGTH 4
+#define AEC_MAX_SOURCES 2 /* Microphone and reference streams */
+
+static const struct comp_driver comp_aec;
+
+/* UUID 2ca424c0-7e1c-4b0a-afca-43de944705d3 */
+DECLARE_SOF_RT_UUID("aec", aec_uuid, 0x2ca424c0, 0x7e1c, 0x4b0a, 0xaf, 0xca,
+		    0x43, 0xde, 0x94, 0x47, 0x05, 0xd3);
+
+DECLARE_TR_CTX(aec_tr, SOF_UUID(aec_uuid), LOG_LEVEL_INFO);
+
+/* Generic processing */
+
+#if CONFIG_FORMAT_S16LE
+
+static void aec_s16_default(struct comp_dev *dev, const struct audio_stream **sources,
+			    struct audio_stream *sink, int frames, int ref_active)
```

Comment:
Also other infomation is needed for reference so better to move ref_active to component data struct.